pkg/operator/ceph/cluster/newosd/osd.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2019 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package newosd
+
+import (
+	"fmt"
+
+	"github.com/coreos/pkg/capnslog"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	rookalpha "github.com/rook/rook/pkg/apis/rook.io/v1alpha2"
+	"github.com/rook/rook/pkg/clusterd"
+	cephconfig "github.com/rook/rook/pkg/daemon/ceph/config"
+	opspec "github.com/rook/rook/pkg/operator/ceph/spec"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "op-osd")
+
+const (
+	appName = "rook-ceph-osd"
+
+	prepareAppName    = "rook-ceph-osd-prepare"
+	prepareAppNameFmt = "rook-ceph-osd-prepare-%s"
+
+	osdAppNameFmt = "rook-ceph-osd-%d"
+
+	osdLabelKey                  = "ceph-osd-id"
+	clusterAvailableSpaceReserve = 0.05
+	serviceAccountName           = "rook-ceph-osd"
+	unknownID                    = -1
+
+	cephOsdPodMinimumMemory uint64 = 4096 // minimum amount of memory in MB to run the pod
+)
+
+// Controller keeps track of the OSDs
+type Controller struct {
+	context     *clusterd.Context
+	namespace   string
+	cephCluster *cephCluster
+	rookImage   string
+	ownerRef    metav1.OwnerReference
+}
+
+type cephCluster struct {
+	spec *cephv1.ClusterSpec
+	info *cephconfig.ClusterInfo
+}
+
+func NewController(
+	context *clusterd.Context,
+	namespace string,
+	cephSpec *cephv1.ClusterSpec,
+	cephInfo *cephconfig.ClusterInfo,
+	ownerRef metav1.OwnerReference,
+) *Controller {
+	return &Controller{
+		context:   context,
+		namespace: namespace,
+		cephCluster: &cephCluster{
+			spec: cephSpec,
+			info: cephInfo,
+		},
+		ownerRef: ownerRef,
+	}
+}
+
+// Start the osd management
+func (c *Controller) Start() error {
+	logger.Infof("OSD controller started for namespace %s", c.namespace)
+
+	// Validate pod's memory if specified
+	r := cephv1.GetOSDResources(c.cephCluster.spec.Resources)
+	err := opspec.CheckPodMemory(r, cephOsdPodMinimumMemory)
+	if err != nil {
+		return fmt.Errorf("user-requested OSD pod resource requirements are insufficient. %+v", err)
+	}
+
+	provisionableNodes, err := c.getProvisionableNodes()
+	if err != nil {
+		return fmt.Errorf("OSD controller failed. %+v", err)
+	}
+	logger.Debugf("returned provisionable nodes: %+v", provisionableNodes)
+
+	// TODO: inventory provisionable nodes
```

Comment:
These are pretty loaded TODOs. :) They look covered by existing implementation, but maybe I shouldn't comment this early in the PR!

Reply:
Since this is a skeleton, I'm expecting to have several TODOs. And much of the logic may be taken from the existing implementation, but it'll still need copied and modified as appropriate.